docker/build/devdeps.Dockerfile

```diff
@@ -18,115 +18,90 @@
-ENV LLVM_INSTALL_PREFIX=/opt/llvm
-
-# Enable compiler-rt in this build (-r)
-RUN source /opt/llvm/bootstrap/init_command.sh && \
-    bash /scripts/build_llvm.sh -s /llvm-project -c Release -v -r \
-    && rm -rf /llvm-project 
-
-# Update bootstrap files to point to new installation
-RUN sed -i "s|/opt/llvm_|/opt/llvm|" /opt/llvm/bootstrap/init_command.sh
-RUN sed -i "s|/opt/llvm_|/opt/llvm|" /opt/llvm/bootstrap/install_toolchain.sh
-
-FROM llvmbuild as prereqs
-ADD ./scripts/install_prerequisites.sh /scripts/install_prerequisites.sh
-RUN apt-get update && apt-get install -y --no-install-recommends ca-certificates \
-    && export LLVM_INSTALL_PREFIX=/opt/llvm \
-    && export BLAS_INSTALL_PREFIX=/usr/local/blas \
-    && export ZLIB_INSTALL_PREFIX=/usr/local/zlib \
-    && export OPENSSL_INSTALL_PREFIX=/usr/local/openssl \
-    && export CURL_INSTALL_PREFIX=/usr/local/curl \
-    # It would be nice to also build the prerequisites
-    # using the same toolchain as CUDA-Q by default.
-    && source /opt/llvm/bootstrap/init_command.sh \
-    && bash /scripts/install_prerequisites.sh \
-    && apt-get remove -y ca-certificates \
-    && apt-get autoremove -y --purge && apt-get clean && rm -rf /var/lib/apt/lists/*
-
-# Pre-built binaries for doxygen are only available for x86_64.
-FROM ${base_image} as doxygenbuild
-RUN if [ "$(uname -m)" != "x86_64" ]; then \
-        apt-get update && apt-get install -y wget unzip make cmake flex bison gcc g++ python3 \
-        # Fixed commit corresponding to release 1.9.7
-        && wget https://github.com/doxygen/doxygen/archive/6a2ce4d18b5af1ca501bcf585e4c8e2b2b353b0f.zip -q -O repo.zip \
-        && unzip repo.zip && mv doxygen* repo && rm repo.zip \
-        && cmake -G "Unix Makefiles" repo && cmake --build . --target install --config Release \
-        && rm -rf repo && apt-get remove -y wget unzip make cmake flex bison gcc g++ python3 \
-        && apt-get autoremove -y --purge && apt-get clean && rm -rf /var/lib/apt/lists/*; \
+        wget git unzip \
+        python3-dev python3-pip && \
+    python3 -m pip install --no-cache-dir numpy && \
+    apt-get autoremove -y --purge && apt-get clean && rm -rf /var/lib/apt/lists/*
+ADD scripts/install_toolchain.sh /cuda-quantum/scripts/install_toolchain.sh
+RUN source /cuda-quantum/scripts/install_toolchain.sh \
+        -e "$LLVM_INSTALL_PREFIX/bootstrap" -t ${toolchain}
+
+## [Source Dependencies]
+ADD scripts/install_prerequisites.sh /cuda-quantum/scripts/install_prerequisites.sh
+ADD scripts/build_llvm.sh /cuda-quantum/scripts/build_llvm.sh
+ADD cmake/caches/LLVM.cmake /cuda-quantum/cmake/caches/LLVM.cmake
+ADD tpls/customizations/llvm /cuda-quantum/tpls/customizations/llvm
+ADD .gitmodules /cuda-quantum/.gitmodules
+ADD .git/modules/tpls/pybind11/HEAD /.git_modules/tpls/pybind11/HEAD
+ADD .git/modules/tpls/llvm/HEAD /.git_modules/tpls/llvm/HEAD
+
+# This is a hack so that we do not need to rebuild the prerequisites 
```

Comment:
How do we make this "The Way" instead of a hack? I know we have bugs in the dependencies, etc. but there isn't a need to rebuild most of the binaries with every CI commit.

Reply:
The comment was copied over from the assets where I did this so that we can validate that the build scripts correctly pick up the commits based on what is checked out in the submodules. For Docker, this initializes the git index sufficiently to pick up submodule commits without actually copying over all the submodules (similarly to the state it would be in if you check out a repo from scratch). I updated the comment, which was probably misleading. We definitively only ever build the prerequisites when something changed, and not in every pipeline.